### PR TITLE
Nystrom cookbook

### PR DESCRIPTION
**Pull request by cfjhallgren**
Tested and rendered html.

### CODE REVIEW

**Comment by cfjhallgren** [comment]
Hmm still not working...

File: src/interfaces/modular/Regression.i [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 /* Remove C Prefix */
 %rename(Regression) CRegression;
+%rename(KRRNystrom) CKRRNystrom;
```
Comment (karlnapf) [inline src/interfaces/modular/Regression.i]: can you send a separate patch for those?

**Comment by karlnapf** [comment]
http://buildbot.shogun-toolbox.org/cookbook_pr/8574a1c30f7b78a2b5ac76318a817c7ec3146788/

**Comment by karlnapf** [comment]
The second check always renders the cookbook for you

**Comment by karlnapf** [comment]
Travis is failing due to interface problems. The SWIG changes you need are two files. Regression.i and Regression_includes.i

**Comment by karlnapf** [comment]
I am away for quite a while now so leave merging this to somebody else (ping ppl in irc)

**Comment by vigsterkr** [comment]
@cfjhallgren lemme know if you need help how to fix the problem that is indicated by travis

File: examples/meta/src/regression/krr_nystrom.sg [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+#![create_features]
+RealFeatures features_train(f_feats_train)
+RealFeatures features_test(f_feats_test)
+RegressionLabels labels_train(f_labels_train)
+RegressionLabels labels_test(f_labels_test)
+#![create_features]
+
+#![create_appropriate_kernel]
+real width = 1
```
Comment (sorig) [inline examples/meta/src/regression/krr_nystrom.sg]: Please use `1.0`. We currently use the `auto` type for C++ translations, so we need to make sure the literals are correctly typed to get the correct serialization for integration testing.

Comment (cfjhallgren) [inline examples/meta/src/regression/krr_nystrom.sg]: Thanks I changed it.

**Comment by cfjhallgren** [comment]
Thanks for your comments. I added the class to Regression.i and Regression_includes.i but still having some issues. The class is available in python but inherited members are missing.

**Comment by cfjhallgren** [comment]
I will continue investigating.

**Comment by cfjhallgren** [comment]
I moved the SWIG updates to a separate patch, after reordering the code which fixed the previous issues. Hopefully this one will succeed after that has been merged.

File: doc/cookbook/source/references.bib [hunk 3]
```diff
@@ -1,3 +1,18 @@
+@inproceedings{williams2001using,
+  title={Using the Nystr{\"o}m method to speed up kernel machines},
+  author={Williams, Christopher and Seeger, Matthias},
```
Comment (karlnapf) [inline doc/cookbook/source/references.bib]: we abbreviate first names in our bibtex

**Comment by karlnapf** [comment]
Can you squash the commits?

File: doc/cookbook/source/examples/regression/krr_nystrom.rst [hunk 4]
```diff
@@ -0,0 +1,69 @@
+===============================
+Nyström Kernel Ridge Regression
+===============================
+
+The Nyström method is a technique for increasing the performance of kernel methods
```
Comment (karlnapf) [inline doc/cookbook/source/examples/regression/krr_nystrom.rst]: "reducing computational load"

File: doc/cookbook/source/examples/regression/krr_nystrom.rst [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+by calculating a lower rank approximation to the kernel matrix. It originally
```
Comment (karlnapf) [inline doc/cookbook/source/examples/regression/krr_nystrom.rst]: via replacing the kernel matrix with a low rank approximation

File: doc/cookbook/source/examples/regression/krr_nystrom.rst [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+appeared in 1930 in the approximation of integral equations.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/regression/krr_nystrom.rst]: Maybe remove the last sentence, not related to Shogun :)

File: doc/cookbook/source/examples/regression/krr_nystrom.rst [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+The approximation is achieved by projecting the data matrix on a subset of data
+points, resulting in a linear system that is cheaper to solve. When applied to
+ridge regression, the following approximate system is solved
+
+.. math::
+    {\bf \alpha} = (\tau K_{m,m} + K_{m,n}K_{n,m})^+K_{m,n} {\bf y}
+
+where :math:`K_{n,m}` is a submatrix of the kernel matrix :math:`K`
+containing all n rows and the m columns
```
Comment (karlnapf) [inline doc/cookbook/source/examples/regression/krr_nystrom.rst]: use math for n and m

File: doc/cookbook/source/examples/regression/krr_nystrom.rst [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+corresponding to the m chosen training examples, :math:`K_{m,n}` is its
+transpose and :math:`K_{m,m}` is the submatrix with the m rows and columns
+corresponding to the training examples chosen. :math:`+` indicates the
+Moore-Penrose pseudoinverse. The complexity is :math:`O(m^2n)`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/regression/krr_nystrom.rst]: I think it is O(m^3 + m^2n), and you should also say that this is for computation.
I would also give memory requirements, and then compare both with the original KRR costs in memory and computation

**Comment by cfjhallgren** [comment]
Thanks for your comments :-) Will fix.

**Comment by karlnapf** [comment]
If we add testing data, this can be merged. Wanna do that? :)

**Comment by karlnapf** [comment]
Ill merge for now, feel free to send a test

**Comment by cfjhallgren** [comment]
Thanks! I'll add the testing data. Teaching finished so I have more programming time from now on.
